### ISSUE TITLE
Keyboard key release delayed (first seen in 1.5.30)

### ISSUE THREAD

**Issue by Mershl** [ibus/ibus#2618]
Which distribution and version?: 
Fedora 40

Which desktop environment and version?: 
Gnome 46

Which session type?: 
Wayland

Which application and version?: 
Xwayland Wine Games (mostly Guild Wars 2)

IBus version?: 
1.5.30-rc1

Issue description:
The release of keys is sometimes delayed for about 200-300ms (very noticeable). The issue happens 3-4 times a minute.

The issue was first seen after upgrading to Fedora 40. Booting Fedora 39 does not show the issue on the same system / same input methods.

Steps to reproduce:
1. Start GW2
1. Move your character in one direction at a time (A, then W, then D, then S)
1. Ocassionally the character continues to move in one direction after releasing the key
1. The unintended movement is stopped after about 200-300ms.

Can you reproduce your problem when you restart ibus-daemon? (yes / no): 
The issue is not reproduciable after running `ibus exit; ibus-daemon --xim &`. Only tested for a few minutes. I'll update the issue.

Do you see any errors when you run ibus-daemon with the verbose option?:
(Run `ibus-daemon --xim --verbose &` and look at the output when you encounter your problem.)
No.

Can you reproduce your problem with a new user account instead of the current your account? (yes / no): 
yes

On a side note. Disk write total is quite high for `ibus-x11` on 1.5.30 (don't remember if I saw that before):
![image](https://github.com/ibus/ibus/assets/520251/84db2acc-2a28-4807-90bc-4caa68a44a66)

**Comment by Mershl** [comment]
The issue is not seen on on fresh boots or after killing ibus. The Disk write total is also much lower on fresh boot. I'll try to find an reproducer to get ibus in this state where the delays and the high Disk write total is seen.

**Comment by Mershl** [comment]
Did another round of testing.

1) Once the issue is present, `killall ibus-daemon` fixes the issue.

2) I was unable to reproduce the issue with `IBUS_ENABLE_SYNC_MODE=2` set. 

3) I wasn't able to find a reproducer for the high `Disk write total`. It's ranging from few kB to 500MB for a 10 minute test.

**Comment by Mershl** [comment]
Shortly discussed this on #flatpak. Flatpak includes ibus-libs, but not its own instance of ibus.

**Comment by Calosis** [comment]
I'm not personally experiencing this on Fedora 40, but my friend is. This is a pretty serious issue as it makes using your computer impossible. See: https://gitlab.gnome.org/GNOME/mutter/-/issues/3369

At first, we thought it was a Mutter bug, but killing Ibus "killall ibus-daemon" fixes the issue. He is using an AMD GPU with Mesa, so it's not related to the other keyboard input lag issue as seen with NVIDIA on Mutter 46.

**Comment by Enderteck** [comment]
> At first, we thought it was a Mutter bug, but killing Ibus "killall ibus-daemon" fixes the issue. He is using an AMD GPU with Mesa, so it's not related to the other keyboard input lag issue as seen with NVIDIA on Mutter 46.

I had the same problem on Nvidia, Fedora 40 on Wayland. 

You saved me so much sanity playing video games without my character continuing to go places for half a sec when I don't touch the movement keys.

**Comment by fujiwarat** [comment]
I tried "Goose Goose Duck" with Steam but I cannot reproduce your issue.

I found "Guild Wars 2" with Steam but it requires 66GB.
Seems Steam does not support NTFS and I have NTFS partitions only for the bigger size.

```
chdir "/run/media/fujiwara/DRIVE_H/Steam/steamapps/common/Guild Wars 2"
ERROR: ld.so: object '/run/media/fujiwara/DRIVE_H/Steam/ubuntu12_32/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS32): ignored.
ERROR: ld.so: object '/run/media/fujiwara/DRIVE_H/Steam/ubuntu12_64/gameoverlayrenderer.so' from LD_PRELOAD cannot be preloaded (wrong ELF class: ELFCLASS64): ignored.
Proton: Upgrading prefix from None to 9.0-201 (/run/media/fujiwara/DRIVE_H/Steam/steamapps/compatdata/1284210/)
Traceback (most recent call last):
  File "/run/media/fujiwara/DRIVE_H/Steam/steamapps/common/Proton - Experimental/proton", line 1553, in <module>
    g_session.init_session(sys.argv[1] != "runinprefix")
  File "/run/media/fujiwara/DRIVE_H/Steam/steamapps/common/Proton - Experimental/proton", line 1463, in init_session
    g_compatdata.setup_prefix()
  File "/run/media/fujiwara/DRIVE_H/Steam/steamapps/common/Proton - Experimental/proton", line 796, in setup_prefix
    os.symlink("../drive_c", self.prefix_dir + "/dosdevices/c:")
OSError: [Errno 22] Invalid argument: '../drive_c' -> '/run/media/fujiwara/DRIVE_H/Steam/steamapps/compatdata/1284210/pfx//dosdevices/c:'
Uploaded AppInterfaceStats to Steam
```

https://forum.manjaro.org/t/steam-proton-not-working-on-ntfs-drive/84912
Seems Steam works with BTRFS and I need to know how to convert NTFS to BTRFS.

**Comment by fujiwarat** [comment]
Finally I could install GW2 in Fedora 40 and reproduce the issue with ibus 1.5.30
But I also can reproduce the issue with 1.5.29 and 1.5.27 (1.5.28 has another issue and it shows worse result.)
So I guess the issue has existed for long time?

I prepared 1.5.29 for Fedora 40 and you could test it:
https://copr.fedorainfracloud.org/coprs/fujiwara/ibus/
```
# rpm -e --nodeps ibus ibus-libs ibus-gtk2 ibus-gtk3 ibus-gtk4 ibus-panel ibus-setup ibus-devel
# dnf copr enable fujiwara/ibus
# dnf install ibus
```

**Comment by Knogle** [comment]
Also affected here on Fedora 40 and random games.

**Comment by PorkDevMode** [comment]
Also having this issue, occurs in overwatch would love to see a fix.

**Comment by Enderteck** [comment]
killall ibus-daemon should fix the issue ?

I ahve it as a startup program

**Comment by ther0n** [comment]
Noticing this issue as well. I noticed that when I was having this issue I was copying a 1TB disk image from an Ext4 partition to an NTFS partition on disk in Nautilus/GNOME Files. `killall ibus-daemon` fixes the issue, will test to see if setting `IBUS_ENABLE_SYNC_MODE=2` in `/etc/environment` fixes the issue without having to kill ibus.

**Comment by sebastianblunt** [comment]
I don't have GW2 but I've been seeing something similar in rxvt-unicode since the update to 1.5.30. I traced it to b49f3a4736540f3cf2f7e7b3ac7f965970393c74, it reliably lags/stutters with the `sync()` but not without it.

It only happens with `IBUS_ENABLE_SYNC_MODE=1` not in mode 0 or 2, seemingly because it never hits that codepath in mode 0 or 2.

I wasn't able to replicate the issue in #2560, with or without `sync()`, but I wonder if the call to sync isn't just effectively fixing the issue by adding a delay (hence why debug messages also fix said messages.)

**Comment by fujiwarat** [comment]
@sebastianblunt 

> I don't have GW2 but I've been seeing something similar in rxvt-unicode since the update to 1.5.30.

I need the reproducing steps in rxvt-unicode.

> I wasn't able to replicate the issue in #2560,

You should try libX11 1.8.7 or older at  first to reproduce the issue and it was fixed in libX11 1.8.8 but libX11 1.8.8 causes a SEGV of ibus-x11 and I applied `sync()`.
libX11 1.8.8 had other regressions and reverted the fix in 1.8.9 but I fixed most issues and the issues and #2560 will be fixed in 1.8.10.

**Comment by x5f3759df** [comment]
Have this issue in games, running through proton - Apex Legend, World War Z or ME Andromeda

Using native Steam on Fedora 40 with AMD GPU with Wayland

```killall ibus-daemon``` fixes issue for game session

```
user@fedora ~> rpm -q ibus
ibus-1.5.30-5.fc40.x86_64
```

**Comment by DesertBlizzard** [comment]
> Also having this issue, occurs in overwatch would love to see a fix.

Seconded.  Same problem, killing ibus-daemon clears this up.  US keyboard layout.

**Comment by nhockey** [comment]
I'm experiencing a strange keyboard input issue. Setting the environment variable XMODIFIERS="@im=null" seems to completely resolve the problem, but I'm unsure why, other than it disabling ibus.

The issue is inconsistent. Sometimes, keystrokes are delayed or lost entirely. For example, I might type a series of characters, but only the last few actually register. Additionally, there are instances where keys appear to be held down long after I've released them, leading to unintended actions in games. This happens with both wired and wireless keyboards. Sometimes the input is all out of order.

To illustrate, I once tried to log into a game but my password didn't register. I clicked on the game window thinking maybe it wasn't focused and typed again, and it worked. However, moments later, while playing the game, my character decided to randomly type my password into world chat.

Additionally different system loads seem to have an effect but not consistently. all i know is it happens worse when i have a youtube video playing on one monitor or something AND am playing the game. This isn't required, but does trigger the issue more frequently. also using multiple keys at the same time effects the frequency. like w+d to move forward while turning right will tend to have a noticeable delay after releasing the keys.

i am using Arch (btw) wih ibus 1.5.30-3 using nvidia closed source drivers with normal old x11 (not wayland)

**Comment by Awesomerly** [comment]
I am having this issue as well on Fedora 40. :)

**Comment by waffshappen** [comment]
AMD GPU (Framework 16) here - i have this when recording games from obs-vkcapture and loading wifi, or any other high cpu/gpu load scenario.

The input delay can be seconds in some games, which is absolutely terrible - killing ibus-daemon __completely__ fixes it.

**Comment by nhockey** [comment]
As a follow-up, I have unset the environment variable XMODIFIERS="@im=null" and switched over to using "IBUS_ENABLE_SYNC_MODE=2". I noticed twitchy mouse movements while the game is loading but virtually no keyboard or mouse issues once the game has loaded. Maybe, occasionally, an extremely slight delay once in a while. Since I have no actual need for ibus I might go back to just disabling it. Though I am still experimenting as I would like to find a resolution for people whom rely on ibus. Or, at the very least, create the correct environment to recreate the issue so that somebody more familiar with the ibus code base could figure out what's going on.

So in summary: enabling sync mode 2 mostly resolves the issue and if you are still experiencing issues I would suggest enabling the environment variable for the X server disabling ibus, unless you rely on it.

**Comment by fujiwarat** [comment]
> Have this issue in games, running through proton - Apex Legend, World War Z or ME Andromeda

I think World War Z is not a free software and I won't try it. I could not find ME Andromeda in steam.

I tried Apex Legend but I cannot find any problems.

[Apex with IBus](https://fujiwara.fedorapeople.org/ibus/20240809/apex-w-ibus.mp4)
[Apex without IBus](https://fujiwara.fedorapeople.org/ibus/20240809/apex-wo-ibus.mp4)

I checked wsad keys and space key for jump also works find in the training screen.

You could run `XMODIFIERS=@im=none steam` to disable IBus.
I have no plan to set IBUS_ENABLE_SYNC_MODE=2.

**Comment by x5f3759df** [comment]
>I tried Apex Legend but I cannot find any problems.

This issue doesn't has consistent reproducible steps, keys ghosting/input buffer is kinda random.

>You could run XMODIFIERS=@im=none steam to disable IBus.

Looks like IBUS_ENABLE_SYNC_MODE=2 in my .profile fixes issue for me.

**Comment by fujiwarat** [comment]
> This issue doesn't has consistent reproducible steps, keys ghosting/input buffer is kinda random.

It does not describe your issue. See above my comment.

**Comment by OhNoMoreGit** [comment]
If you want a simpler reproducer than a game, a basic X text editor like Nedit shows the problem clearly. It's practically unusable unless `XMODIFIERS` is cleared.

`IBUS_ENABLE_SYNC_MODE=2` does _not_ seemingly work around the problem. Only clearing `XMODIFIERS` appears to work.

**Comment by Calosis** [comment]
> I'm not personally experiencing this on Fedora 40, but my friend is. This is a pretty serious issue as it makes using your computer impossible. See: https://gitlab.gnome.org/GNOME/mutter/-/issues/3369
> 
> At first, we thought it was a Mutter bug, but killing Ibus "killall ibus-daemon" fixes the issue. He is using an AMD GPU with Mesa, so it's not related to the other keyboard input lag issue as seen with NVIDIA on Mutter 46.

As shown here, in the Mutter issue created, in an IDE like JetBrains IntelliJ the issue is present.

**Comment by fujiwarat** [comment]
Now ibus-1.5.30-6.fc40 and libX11-1.8.10-1.fc40 in Fedora 40, ibus-1.5.31~beta1-12.fc41 and libX11-1.8.10-1.fc41 in Fedora 41 and I'd ask you to test the packages if you still reproduce your issues.
I applied https://github.com/ibus/ibus/pull/2673 to ibus side.

**Comment by Mershl** [comment]
Hi @fujiwarat,
thank you very much! can confirm

| ibus  | libX11 | issue reproduciable on Fedora 40 (Gnome Wayland) |
| ------------- | ------------- | ------------- |
| `1.5.30-5.fc40`  | `1.8.10-1.fc40.x86_64`  | Issue is instantly reproduciable in Guild Wars 2  |
| `1.5.30-6.fc40` | `1.8.10-1.fc40.x86_64` | Issue is no longer reproduciable in any tested game on my machine  |

Koji build of `1.5.30-6.fc40` for anyone interested in testing: https://koji.fedoraproject.org/koji/buildinfo?buildID=2529273

**Comment by fujiwarat** [comment]
I don't push the PR to upstream yet.

**Comment by fujiwarat** [comment]
Let's see if #2673 can fix the issue. I will release 1.5.31-beta2 ASAP.

**Comment by Mershl** [comment]
I'm not yet sure if this is the same issue, but I'm experiencing a similar issue in Fedora 41 again (`ibus-1.5.31~beta2-2.fc41.x86_64`)

**Comment by ofirtim** [comment]
Can confirm still goes on.
Its not only delays, its keys and mouse clicks locking in when running most Proton/Wine games on Heroic/Steam or Lutris.
Doesn't happened on Native games like Counter Strike 2.

Had disabled Ibus and seems to solve the issue.
Also noticed that while that delay was coming after some time of playing, Ibus was using like 4-5% of my CPU, which is very weird.
